pkg/controller/registry/resolver/solver/constraint_provider.go

```diff
@@ -0,0 +1,15 @@
+package solver
+
+import "github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/resolver/cache"
+
+// ConstraintProvider knows how to provide solver constraints for a given cache entry.
+type ConstraintProvider interface {
+	// Constraints returns a set of solver constraints for a cache entry.
+	Constraints(e *cache.Entry) ([]Constraint, error)
+}
+
+type ConstraintProviderFunc func(e *cache.Entry) ([]Constraint, error)
+
+func (c ConstraintProviderFunc) Constraints(e *cache.Entry) ([]Constraint, error) {
```

Comment:
Comment here would be nice to indicate this func extracts constraints from catalog entry (bundle).

Reply:
It's not quite for extracting constraints from the bundle, but providing additional constraints. I've added some comments here and to the ConstraintProvider interface as well

Reply:
This interface is a bit more confusing than I thought but I got what you meant.